lib/spack/spack/build_environment.py

```diff
@@ -1016,10 +1016,17 @@ def get_env_modifications(self) -> EnvironmentModifications:
                 self._make_runnable(dspec, env)
 
             if self.should_setup_run_env & flag:
+                run_env_mods = EnvironmentModifications()
                 for spec in dspec.dependents(deptype=dt.LINK | dt.RUN):
                     if id(spec) in self.nodes_in_subdag:
-                        pkg.setup_dependent_run_environment(env, spec)
-                pkg.setup_run_environment(env)
+                        pkg.setup_dependent_run_environment(run_env_mods, spec)
+                pkg.setup_run_environment(run_env_mods)
+                if self.context == Context.BUILD:
```

Comment:
Add a `TODO` here to handle this better when we have proper "compiler" dependencies, mention this is a stopgap